mii/config.py

```diff
@@ -3,93 +3,147 @@
-    # listed value when Enum is called
-    def __new__(cls, *values):
-        obj = object.__new__(cls)
-        # first value is canonical value
-        obj._value_ = values[0]
-        for other_value in values[1:]:
-            cls._value2member_map_[other_value] = obj
-        obj._all_values = values
-        return obj
-
-    def __repr__(self):
-        return "<%s.%s: %s>" % (
-            self.__class__.__name__,
-            self._name_,
-            ", ".join([repr(v) for v in self._all_values]),
-        )
+import os
+import string
+from typing import List, Optional, Dict, Any
+from pydantic import validator, root_validator
+import mii
+from mii.constants import DeploymentType, TaskType, MII_MODEL_PATH_DEFAULT
 
+from deepspeed.runtime.config_utils import DeepSpeedConfigModel
+from deepspeed.inference.config import DtypeEnum
+from deepspeed.launcher.runner import DLTS_HOSTFILE, fetch_hostfile
 
-class MIIConfig(BaseModel):
-    tensor_parallel: int = 1
-    port_number: int = 50050
-    dtype: DtypeEnum = torch.float32
-    meta_tensor: bool = False
+
+class ReplicaConfig(DeepSpeedConfigModel):
+    hostname: str = ""
+    tensor_parallel_ports: List[int] = []
+    torch_dist_port: int = None
+    gpu_indices: List[int] = []
+
+
+class DeploymentConfig(DeepSpeedConfigModel):
+    # Deployment configs
+    deployment_name: str
     load_with_sys_mem: bool = False
-    enable_cuda_graph: bool = False
-    checkpoint_dict: Union[dict, None] = None
-    deploy_rank: Union[int, List[int]] = -1
+    meta_tensor: bool = False
+    hf_auth_token: str = ""
+    deploy_rank: Optional[List[int]] = None
     torch_dist_port: int = 29500
-    hf_auth_token: str = None
-    replace_with_kernel_inject: bool = True
+    replica_num: int = 1
+    replica_configs: List[ReplicaConfig] = []
     profile_model_time: bool = False
     skip_model_check: bool = False
-    max_tokens: int = 1024
-    enable_restful_api: bool = False
-    restful_api_port: int = 51080
-    replica_num: int = 1
-    hostfile: str = DLTS_HOSTFILE
     trust_remote_code: bool = False
 
-    @validator("deploy_rank")
-    def deploy_valid(cls, field_value, values):
-        if "tensor_parallel" not in values:
-            raise ValueError(
-                "'tensor_parallel' must be defined in the pydantic model before 'deploy_rank'"
-            )
-
-        # if deploy rank is not given, default to align with TP value
-        if field_value == -1:
-            field_value = list(range(values["tensor_parallel"]))
+    # Model configs
+    model: str
+    task: TaskType
+    dtype: DtypeEnum = DtypeEnum.fp32
+    model_path: str = ""
+    checkpoint_dict: Optional[Dict[str, Any]] = None
+    max_tokens: int = 1024
 
-        # ensure deploy rank type is always list for easier consumption later
-        if not isinstance(field_value, list):
-            field_value = [field_value]
+    # Performance configs
```

Comment:
Do we want to use DeepSpeed's `InferenceConfig` here? This would expand on the options that we expose to users and we could pass the config directly to DeepSpeed in `load_models.py`!